datafusion/common/src/column.rs

```diff
@@ -18,22 +18,35 @@
 //! Column
 
 use arrow_schema::{Field, FieldRef};
+use sqlparser::tokenizer::Span;
 
 use crate::error::_schema_err;
 use crate::utils::{parse_identifiers_normalized, quote_identifier};
 use crate::{DFSchema, Result, SchemaError, TableReference};
+use derivative::Derivative;
 use std::collections::HashSet;
 use std::convert::Infallible;
 use std::fmt;
 use std::str::FromStr;
 
 /// A named reference to a qualified field in a schema.
-#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+#[derive(Debug, Clone, Derivative)]
+#[derivative(PartialEq, Eq, Hash, PartialOrd, Ord)]
 pub struct Column {
     /// relation/table reference.
     pub relation: Option<TableReference>,
     /// field/column name.
     pub name: String,
+    #[derivative(
```

Comment:
this is an interesting pattern -- it is different than the way sqlparser did it (which is to effectively wrap the span with a struct that is ignored when comparing eq, hash, etc

Did you consider a similar approach? 

```rust
pub struct Column {
...
  // Wrapped span that does not contribute to PartialEq, Eq, Hash, etc
  pub span: DiagnosticOnly<Span>
}
```

I think this approach is less verbose and might be easier to understand for the casual reader

Reply:
I also see `WithSpan` has a similar approach (but that is for wrapping things with Spans 🤔 )

Reply:
It is indeed interesting. 😅 I was going to use derivative for this problem in sqlparser initially actually, but I wasn't sure if you would want to add a new dependency. If that is not a concern, I think using derivative for cases like this is pretty nice as it is also clear at the usage site that it is being ignored, unlike `AttachedToken` which kind of hides the behavior. If a wrapper type is preferred, I now think actually maybe the struct in sqlparser should have been called DiagnosticOnly as well 😂

Reply:
I think the [`AttachedToken`](https://github.com/apache/datafusion-sqlparser-rs/blob/main/src/ast/helpers/attached_token.rs) approach that we took in `sqlparser` could be a functionally equivalent alternative. And I understand that not everybody might be familiar with `derivative`.

What I don't 100% love about it is that the `PartialEq` implementation on `AttachedToken` is a bit "impure". In that, it serves the one specific purpose of making it ignored when used in a struct field, but prevents you from actually comparing two instances of it because they would always are equal (the implementation returns always `true`), which is something you might want to do at some point. Especially because the name `AttachedToken` doesn't clearly convey that intent, and does more than just being a passthrough for `PartialEq` since its main purpose is tying together a `Span` and a `Token`.

When looking at a struct that uses it:

```rust
#[derive(Debug, Clone, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub struct Select {
    /// Token for the `SELECT` keyword
    pub select_token: AttachedToken,
	// ...
}
```

it's not clear to me that `AttachedToken` is being ignored. I only realise that when I  go look at the implementation of `PartialEq`.

Overall, the `derivative` approach to me more clearly lets me keep the sensible implementation of `PartialEq` for `Span`, while also conveying that "when used as a field in this particular struct, I want it to not influence the struct's comparison". But in other structs, I might want it to influence the comparison.

I personally find this approach more readable and flexible, but I'm open to converting to a wrapper type :)

Reply:
> I also see WithSpan has a similar approach (but that is for wrapping things with Spans 🤔 )

Ah yes good catch! I could definitely have used `derivative` for `WithSpan` too. And about `WithSpan`: do you like the general idea of it?

It's meant to be a "sidecar" to get a `Span` all the way to where it's needed to create a meaningful diagnostic, by attaching it to values that are sort-of-related, but wouldn't make a lot of sense to add as a struct field.

For example, the two `DataType`s in a binary expression come from the two expressions in the SQL query, so they (in some sense) relate to a portion of the code. But putting `span` as a field of `DataType` sounds wrong.

Maybe a better name could be `SpanRelated` or something like that?.

Reply:
From my perspective some design goals should be:
1. Make it easy for someone reading the DataFusion source code to understand what is going on
2. If people are not interested in the span / diagnostic information they can ignore it

From my perspective, using new crates does reduce the amount of code in the DataFusion crate, but can often increase cognitive load (as now to understand the DataFusion code you need to understand the crates)

> I think the AttachedToken approach that we took in sqlparser could be a functionally equivalent alternative. And I understand that not everybody might be familiar with derivative.

Yeah, that is why I think AttachedToken might be better in this case